### PR TITLE
feat(common): remove null check in deserialization method for optional properties of non-mgmt sdk

### PR DESCRIPTION
**Pull request by archerzz**
- for models of non-mgmt sdk, do not generate null check logic for optional properties in deserialization method
- change the null value of `Optional<T>` from `null` to `default`, so that we can assign null value to value types
- add test cases for `Optional<T>` regarding null value
- minor refactoring
- update test/samples

# Description
<i>Add your description here!</i>

# Checklist

To ensure a quick review and merge, please ensure:
- [ ] The PR has a understandable title and description explaining the _why_ and _what_.
- [ ] The PR is opened in draft if not ready for review yet.
   - If opened in draft, please allocate sufficient time (24 hours) after moving out of draft for review
- [ ] The branch is recent enough to not have merge conflicts upon creation.

# Ready to Land?
- [ ] Build is completely green
   - Submissions with test failures require tracking issue and approval of a CODEOWNER
- [ ] At least one +1 review by a CODEOWNER
- [ ] All -1 reviews are confirmed resolved by the reviewer 
   - Override/Marking reviews stale must be discussed with CODEOWNERS first

### CODE REVIEW

File: src/AutoRest.CSharp/Common/Generation/Writers/JsonCodeWriterExtensions.cs [hunk 1]
```diff
@@ -304,11 +304,12 @@ private static void DeserializeIntoObjectProperties(this CodeWriter writer, IEnu
...
                              true) //https://github.com/Azure/autorest.csharp/issues/922
                     {
-                        if (Configuration.AzureArm && property.ValueType?.Equals(typeof(Uri)) == true)
+                        if ((!Configuration.AzureArm || property.ValueType?.Equals(typeof(Uri)) == true)
+                            && !property.IsFlattenedProperty) // if flatten property, then we should not try to assign null/default to parent property
                         {
                             using (writer.Scope($"if ({itemVariable}.Value.ValueKind == {typeof(JsonValueKind)}.Null)"))
                             {
-                                writer.Line($"{propertyVariables[property].Declaration} = null;");
+                                writer.Line($"{propertyVariables[property].Declaration} = default;");
```
Comment (ArcturusZhang) [inline src/AutoRest.CSharp/Common/Generation/Writers/JsonCodeWriterExtensions.cs:312]: is it possible that we do not change this? Or add a check if the property is reference type we still use null instead of default to reduce the amount of changed code?

File: src/AutoRest.CSharp/Common/Generation/Writers/JsonCodeWriterExtensions.cs [hunk 2]
```diff
@@ -329,13 +330,13 @@ private static void DeserializeIntoObjectProperties(this CodeWriter writer, IEnu
                         var variableOrExpression = writer.DeserializeValue(property.ValueSerialization, $"{itemVariable}.Value");
                         writer.Line($"{propertyVariables[property].Declaration} = {variableOrExpression};");
                     }
-                    else if (property.PropertySerializations is not null)
+                    else if (property.IsFlattenedProperty)
```
Comment (ArcturusZhang) [inline src/AutoRest.CSharp/Common/Generation/Writers/JsonCodeWriterExtensions.cs:333]: Well this name collides with something I introduced for the "safe flatten" in mgmt plane.
Shall we rename this to `IsClientFlattenedProperty` to specify?

**Comment by m-nash** [comment]
@archerzz  we may have discussed this already but help me understand why we want to exclude mgmt plane from this?  @ArthurMa1978 I thought we discussed doing something similar to this for mgmt plane due to all the `x-nullable` flags we have had to manually add to autorest.md to fix incomplete specs.

**Comment by archerzz** [comment]
> @archerzz we may have discussed this already but help me understand why we want to exclude mgmt plane from this? @ArthurMa1978 I thought we discussed doing something similar to this for mgmt plane due to all the `x-nullable` flags we have had to manually add to autorest.md to fix incomplete specs.

@m-nash It's to minimize the impact scope, so we excluded the change to mgmt plane.

**Comment by archerzz** [comment]
Preview on `azure-sdk-for-net`: https://github.com/Azure/azure-sdk-for-net/pull/34630

**Comment by archerzz** [comment]
I've filed an issue on swagger to add null check in live validation: https://github.com/Azure/azure-rest-api-specs/issues/23217

File: src/AutoRest.CSharp/Common/Generation/Writers/JsonCodeWriterExtensions.cs [hunk 3]
```diff
@@ -361,6 +360,13 @@ private static void DeserializeIntoObjectProperties(this CodeWriter writer, IEnu
             }
         }
 
+        private static FormattableString GetNewCollectionExpression(JsonSerialization serialization) => serialization switch
```
Comment (AlexanderSher) [inline src/AutoRest.CSharp/Common/Generation/Writers/JsonCodeWriterExtensions.cs:363]: Is it used anywhere?

File: src/AutoRest.CSharp/Common/Generation/Writers/JsonCodeWriterExtensions.cs [hunk 4]
```diff
@@ -317,11 +317,10 @@ private static void DeserializeIntoObjectProperties(this CodeWriter writer, IEnu
                              property.ValueType?.Equals(typeof(string)) !=
                              true) //https://github.com/Azure/autorest.csharp/issues/922
                     {
-                        if (Configuration.AzureArm && property.ValueType?.Equals(typeof(Uri)) == true)
+                        if (!property.IsClientFlattenedProperty) // if flatten property, then we should not try to assign null/default to parent property
```
Comment (AlexanderSher) [inline src/AutoRest.CSharp/Common/Generation/Writers/JsonCodeWriterExtensions.cs:320]: It is better to use `property.ValueSerialization != null` here. That way it has meaning - value of the property can't be deserialized directly (reason is not important).

Comment (archerzz) [reply src/AutoRest.CSharp/Common/Generation/Writers/JsonCodeWriterExtensions.cs]: expanded the method

File: src/AutoRest.CSharp/Common/Generation/Writers/JsonCodeWriterExtensions.cs [hunk 5]
```diff
@@ -328,11 +328,10 @@ private static void DeserializeIntoObjectProperties(this CodeWriter writer, IEnu
                              true) //https://github.com/Azure/autorest.csharp/issues/922
                     {
                         var emptyStringCheck = GetEmptyStringCheckClause(property, itemVariable, shouldTreatEmptyStringAsNull);
-                        if (Configuration.AzureArm && property.ValueType?.Equals(typeof(Uri)) == true)
+                        if (property.PropertySerializations is null)
```
Comment (m-nash) [inline src/AutoRest.CSharp/Common/Generation/Writers/JsonCodeWriterExtensions.cs:331]: nit: any reason we don't use `!property.IsClientFlattenedProperty`?

**Review by m-nash — APPROVED** [review]

File: src/AutoRest.CSharp/Common/Generation/Writers/JsonCodeWriterExtensions.cs (hunk 5 above)
Comment (ArcturusZhang) [reply src/AutoRest.CSharp/Common/Generation/Writers/JsonCodeWriterExtensions.cs:330]: In the meantime I moved the comment on `IsClientFlattenedProperty` to `PropertySerializations` so that one could know what this property is for

Comment (archerzz) [reply src/AutoRest.CSharp/Common/Generation/Writers/JsonCodeWriterExtensions.cs:330]: This is per Alexander's comment. `PropertySerialization is null` might be better here.
